### PR TITLE
Initial pass at code review document

### PR DESCRIPTION
**Pull request by goldenapples**

### CODE REVIEW

**Comment by goldenapples** [comment]
Addresses #18.

**Comment by danielbachhuber** [comment]
@fusioneng/fusion-engineering Can you give this a read by EOD Tuesday and leave any comments or questions you might have?

File: team-culture/code-review.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+- We're currently using Github for everything. Read up on [our Github
+  workflow](../tools/github.md) for details of process, naming, usage of issues
+  and pull requests.
+- A moderately complex issue will typically be addressed in multiple pull
+  requests, each tackling a distinct part of the issue. 
+- On every commit pushed to a GH branch, CI automated tests are run on Travis.
+  If the build passes, the PR can be reviewed. If not, the original developer
+  will receive email notification that their pull needs to be fixed before it
+  can be reviewed or merged.
```
Comment (danielbachhuber) [inline team-culture/code-review.md]: What CI automated tests are these?

File: team-culture/code-review.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
+# Code Review
+
+A constant process of code review is necesary to maintain quality of code
+output, as well as building cohesive style that can make all our work easier to
```
Comment (danielbachhuber) [inline team-culture/code-review.md]: > maintain quality of code output [...] cohesive style that can make all our work easier

Why are these important?

File: team-culture/code-review.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  and pull requests.
+- A moderately complex issue will typically be addressed in multiple pull
+  requests, each tackling a distinct part of the issue. 
+- On every commit pushed to a GH branch, CI automated tests are run on Travis.
+  If the build passes, the PR can be reviewed. If not, the original developer
+  will receive email notification that their pull needs to be fixed before it
+  can be reviewed or merged.
+- When the original developer is satisfied with their work, they can assign the
+  pull request to another team member for review. Simple pull requests can be
+  merged by the developer who reviews them.
```
Comment (danielbachhuber) [inline team-culture/code-review.md]: I think it's important to underscore the importance of each developer code reviewing their own work first, before passing it off. Each developer needs to verify the code changes expressed in their diff.

File: team-culture/code-review.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+A good way to visualize the objectives of code review is this analogy to
+Maslow's ["Hierarchy of
+Needs"](http://blog.d3in.org/post/111338685456/maslows-pyramid-of-code-review)
+pyramid. As a reviewer, you will be looking at the code on many levels. But its
+important to keep a sense of priorities: if a change introduces unhandled edge
+cases, bugs, or security vulnerabilities, those issues need to be addressed
+before coding style guidelines or beautification practices preferences will
+matter.
+
+![code-review-pyramid](https://cloud.githubusercontent.com/assets/665992/7326019/1603bcd2-ea77-11e4-8510-4d7f76ca2ad1.png)
```
Comment (danielbachhuber) [inline team-culture/code-review.md]: Can you include a quick summary of the hierarchy for those who don't click through?

File: team-culture/code-review.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
+# Code Review
+
+A constant process of code review is necesary to maintain quality of code
+output, as well as building cohesive style that can make all our work easier to
+approach and maintain. If done right, code review can also be a learning
+process for all involved. Here's how we do it at Fusion:
```
Comment (danielbachhuber) [inline team-culture/code-review.md]: Would be nice to have relative links to each subsection.

File: team-culture/code-review.md (hunk 1 above)
Comment (goldenapples) [inline team-culture/code-review.md]: Its more of an aspirational thing at this point. There ideally *should* be behavioral & acceptance tests as well as unit tests. Listed a few in f318ea9.

File: team-culture/code-review.md (hunk 2 above)
Comment (goldenapples) [inline team-culture/code-review.md]: Tried to address in f963152; I'm sure there's a better explanation I can find and link to.

File: team-culture/code-review.md (hunk 4 above)
Comment (goldenapples) [inline team-culture/code-review.md]: f08072d

File: team-culture/code-review.md (hunk 5 above)
Comment (goldenapples) [inline team-culture/code-review.md]: f963152

**Comment by lisaej** [comment]
At the moment:
- We have Matt and Keith doing QA - should such QA/Product manager roles be mentioned in the doc?
- Isn't it only Daniel who does merging? Would there be other reviewer/mergers?

:+1:  to @goldenapples

**Comment by goldenapples** [comment]
> should such QA/Product manager roles be mentioned in the doc?

Good point. I was tackling code reviews in isolation, but really they probably need to be contextualized as one step in the overall QA/approval process.

> Isn't it only Daniel who does merging? Would there be other reviewer/mergers?

At the moment, yes. But I don't think that's scalable. Maybe right now nothing is being held up waiting on review/merges, but as the team grows, being able to review each others' work would eliminate some potential blockage. And even if pulls eventually go through Daniel for final merge, there's a lot of value in peer reviewing proposed changes first.

File: team-culture/code-review.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+We're currently using Github for everything. Read up on [our Github
+workflow](../tools/github.md) for details of process, naming, usage of issues
+and pull requests. A moderately complex issue will typically be addressed
+across multiple pull requests, each tackling a distinct part of the issue. This
+makes review simpler, because each review stage wil be looking at a small code.
+
+- On every commit pushed to a GH branch, our CI automated tests -- linting for
+  all code, unit tests for functional code, and ideally, behavioral and
+  sutomated acceptance testing -- are run on Travis.  If the build passes,
```
Comment (noppanit) [inline team-culture/code-review.md]: Typo? `sutomated`

File: team-culture/code-review.md [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+  the pull request to a team lead to merge and push to production.
+
+
+## What code review is
+
+A good way to visualize the objectives of code review is [this
+analogy](http://blog.d3in.org/post/111338685456/maslows-pyramid-of-code-review)
+to Maslow's "Hierarchy of Needs" pyramid. From most basic to the highest level,
+a reviewer is checking that code is **Correct**, **Secure**, **Readable**,
+**Elegant**, **Elegant**, and **Altruistic**. It's important to keep this sense
```
Comment (noppanit) [inline team-culture/code-review.md]: `+**Elegant**, **Elegant**, `

double Elegant.

File: team-culture/code-review.md [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+- [How to review code](#how-to-review-code)
+- [How to receive code review](#how-to-receive-code-review)
+
+## A typical code review workflow
+
+We're currently using Github for everything. Read up on [our Github
+workflow](../tools/github.md) for details of process, naming, usage of issues
+and pull requests. A moderately complex issue will typically be addressed
+across multiple pull requests, each tackling a distinct part of the issue. This
+makes review simpler, because each review stage wil be looking at a small code.
```
Comment (mpatek) [inline team-culture/code-review.md]: looking at a small change to the code.

**Comment by danielbachhuber** [comment]
> Maybe right now nothing is being held up waiting on review/merges, but as the team grows, being able to review each others' work would eliminate some potential blockage. And even if pulls eventually go through Daniel for final merge, there's a lot of value in peer reviewing proposed changes first.

Once we feel comfortable with our code review routines, and it's clear the team is willing to own the quality of what they deploy to production, [we'll set up some auto-deploy mechanism with Travis](https://trello.com/c/lcTE1wtQ/95-auto-deploy-master-branch-to-vip-from-travis) so master branch is automatically pushed.

File: team-culture/code-review.md [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+and known patterns, so that others can understand it at a glance. If a change
+introduces a chance to refactor surrounding functionality, to abstract and
+standardize old code into new patterns, suggest those opportunities. 
+
+
+### How to receive code review
+
+*As the person receiving the code review*, your job is to learn from
+suggestions. Defensiveness, stubbornness, or impatience can prevent you from
+getting the most out of suggestions. Its OK to explain yourself if you feel the
```
Comment (mpatek) [inline team-culture/code-review.md]: It's

File: team-culture/code-review.md [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+guidelines or beautification practices preferences will matter.
+
+![code-review-pyramid](https://cloud.githubusercontent.com/assets/665992/7326019/1603bcd2-ea77-11e4-8510-4d7f76ca2ad1.png)
+
+
+### How to review code
+
+*As a reviewer*, your first job is to get an understanding of what the proposed
+change does and why it's essential. There's no point in critiquing anything
+until you understand what it does, why its necessary, and what decisions went
```
Comment (FStop) [inline team-culture/code-review.md]: it's

**Pull request merged**